activerecord/lib/active_record/autosave_association.rb

```diff
@@ -521,6 +521,16 @@ def save_belongs_to_association(reflection)
         end
       end
 
+      def compute_primary_key(reflection, record)
+        if primary_key_options = reflection.options[:primary_key]
+          primary_key_options
+        elsif query_constraints = record.class.query_constraints_list
+          query_constraints
+        else
+          :id
```

Comment:
This `:id` is a reimplementation of the previous feature here https://github.com/rails/rails/blob/03702da57aa4e607e0115ea063908f9d4df7c3a0/activerecord/lib/active_record/autosave_association.rb#L508.

I am surprised we were hardcoding `:id` like this and after we merge this I think we should consider getting this from elsewhere. I feel like there's an old bug here since it's not using primary key from the record or reflection.

Reply:
I had the same concern and even tried to switch to `reflection.active_record_primary_key`, which technically was supposed to be equal to calling `record.id` but it didn't work because of at least one test failing so I decided not to dig into this to avoid unnecessary breaking things - https://github.com/rails/rails/pull/47230#discussion_r1094665781

Though design-wise I still see `reflection` to be a more appropriate source of the primary key columns instead of a hardcoded `:id` method call, which in Rails refers to model's `@primary_key`